otherlibs/systhreads/Makefile

```diff
@@ -13,96 +13,127 @@
+MKLIB=$(CAMLRUN) $(ROOTDIR)/tools/ocamlmklib
 COMPFLAGS=-w +33..39 -warn-error A -g -bin-annot -safe-string
 ifeq "$(FLAMBDA)" "true"
 OPTCOMPFLAGS=-O3
 else
 OPTCOMPFLAGS=
 endif
 
-BYTECODE_C_OBJS=st_stubs_b.o
-NATIVECODE_C_OBJS=st_stubs_n.o
+LIBNAME=threads
+
+ifeq "$(UNIX_OR_WIN32)" "unix"
+HEADER = st_posix.h
+else # Windows
+HEADER = st_win32.h
+endif
+
+BYTECODE_C_OBJS=st_stubs_b.$(O)
+NATIVECODE_C_OBJS=st_stubs_n.$(O)
+
+THREADS_SOURCES := thread.ml mutex.ml condition.ml event.ml threadUnix.ml
 
-THREAD_OBJS= thread.cmo mutex.cmo condition.cmo event.cmo threadUnix.cmo
+THREADS_BCOBJS := $(THREADS_SOURCES:.ml=.cmo)
+THREADS_NCOBJS := $(THREADS_SOURCES:.ml=.cmx)
 
-all: libthreads.a threads.cma
+MLIFILES=thread.mli mutex.mli condition.mli event.mli threadUnix.mli
+CMIFILES=$(MLIFILES:.mli=.cmi)
 
-allopt: libthreadsnat.a threads.cmxa
+all: lib$(LIBNAME).$(A) $(LIBNAME).cma $(CMIFILES)
 
-libthreads.a: $(BYTECODE_C_OBJS)
-	$(MKLIB) -o threads $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
+allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
 
-st_stubs_b.o: st_stubs.c st_posix.h
-	$(BYTECC) -I../../byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) \
-	   -c st_stubs.c
-	mv st_stubs.o st_stubs_b.o
+lib$(LIBNAME).$(A): $(BYTECODE_C_OBJS)
+	$(MKLIB) -o $(LIBNAME) $(BYTECODE_C_OBJS) $(PTHREAD_LINK)
 
+lib$(LIBNAME)nat.$(A): $(NATIVECODE_C_OBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
 # Dynamic linking with -lpthread is risky on many platforms, so
 # do not create a shared object for libthreadsnat.
-libthreadsnat.a: $(NATIVECODE_C_OBJS)
-	$(AR) rc libthreadsnat.a $(NATIVECODE_C_OBJS)
+	$(AR) rc $@ $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
+endif
 
-st_stubs_n.o: st_stubs.c st_posix.h
-	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
-	            $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
-	            -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
-	mv st_stubs.o st_stubs_n.o
+$(LIBNAME).cma: $(THREADS_BCOBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
+endif
 
-threads.cma: $(THREAD_OBJS)
-	$(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
-	  -cclib -lunix $(PTHREAD_CAML_LINK)
+# TODO: figure out why linkall is used only for Windows
 
 # See remark above: force static linking of libthreadsnat.a
-threads.cmxa: $(THREAD_OBJS:.cmo=.cmx)
-	$(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
-	  -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+$(LIBNAME).cmxa: $(THREADS_NCOBJS)
+	$(CAMLOPT) -a -o $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
 
 # Note: I removed "-cclib -lunix" from the line above.
 # Indeed, if we link threads.cmxa, then we must also link unix.cmxa,
 # which itself will pass -lunix to the C linker.  It seems more
 # modular to me this way. -- Alain
 
+$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
```

Comment:
Alain Frisch (2016/08/30 14:47 -0700):
> The .cmxs file was not built on Unix previously, probably because
> "Dynamic linking with -lpthread is risky on many platforms" as another
> comment said.  It used to be built under Windows, though.  I'd rather
> not starting to build the .cmxs file under Unix, unless you have a
> good reason to do so.  And I would use the opportunity to stop
> building it under Windows as well for the sake of uniformity across
> platforms.

I think cmxs files are provided for other libraries, see for instance
the rules in otherlibs/Makefile. So my (open) question is: is it okay to
not be consistent in providing the .cmxs files?

I mean, as I understand it, there is a choice between either not
providing the .cmxs file for this library, for sake of obtaining a
simpler build system, or trying to provide the .cmxs file also for this
library, whenever possible (i.e. at least on Windows) because this is
more coherent with what happens with other libraries.

Is it correct to phrase the question that way?

I, personnally, have no strong opinion. It seems to me that the overall
coherence (providing .cmxs files whenever possible) could be considered as
more important than keeping the build system uniform, because it can
have an impact on users, whereas only OCaml developers are concerned
with the simplicity of the build system. But that beiing said I'm happy
to implement whatever the community finds best.